core/src/main/java/com/microsoft/applicationinsights/channel/concrete/inprocess/ThrottlingHandler.java

```diff
@@ -0,0 +1,137 @@
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+
+/**
+ * This class implements the retry logic for throttled requests. HTTP status
+ * code 429 and 439.
+ * 
+ * @author jamdavi
+ *
+ */
+public class ThrottlingHandler implements TransmissionHandler {
+
+	private TransmissionPolicyManager transmissionPolicyManager;
+	private final static String RESPONSE_RETRY_AFTER_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss";
+
+	/**
+	 * Ctor
+	 * 
+	 * Constructs the ThrottlingHandler object.
+	 * 
+	 * @param policy
+	 *            The {@link TransmissionPolicyManager} object that is needed to
+	 *            control the back off policy.
+	 */
+	public ThrottlingHandler(TransmissionPolicyManager policy) {
+		this.transmissionPolicyManager = policy;
+	}
+
+	@Override
+	public void onTransmissionSent(TransmissionHandlerArgs args) {
+		validateTransmissionAndSend(args);
+	}
+
+	/**
+	 * Provides the core logic for the retransmission
+	 * 
+	 * @param args
+	 *            The {@link TransmissionHandlerArgs} for this transmission.
+	 * @return Returns a pass/fail for handling this transmission.
+	 */
+	boolean validateTransmissionAndSend(TransmissionHandlerArgs args) {
+		if (args.getRetryHeader() != null && args.getTransmission() != null
+				&& args.getTransmissionDispatcher() != null) {
+			args.getTransmission().incrementNumberOfSends();
+			switch (args.getResponseCode()) {
+			case 429:
+			case 439:
+				suspendTransmissions(TransmissionPolicy.BLOCKED_BUT_CAN_BE_PERSISTED, args.getRetryHeader());
+				args.getTransmissionDispatcher().dispatch(args.getTransmission());
+				return true;
+			default:
+				InternalLogger.INSTANCE.trace("Http response code %s not handled by %s", args.getResponseCode(),
+						this.getClass().getName());
+				return false;
+			}
+		}
+		InternalLogger.INSTANCE.trace("Http response code %s not handled by %s.", args.getResponseCode(),
+				this.getClass().getName());
+		return false;
+	}
+
+	/**
+	 * Used to put the sender thread to sleep for the specified duration in the
+	 * Retry-After header.
+	 * 
+	 * @param suspensionPolicy
+	 *            The policy used to suspend the threads. For now we use
+	 *            {@link TransmissionPolicy.BLOCKED_BUT_CAN_BE_PERSISTED} for reuse
+	 *            of the existing logic.
+	 * @param retryAfterHeader
+	 *            The header that is captured from the HTTP response.
+	 */
+	private void suspendTransmissions(TransmissionPolicy suspensionPolicy, Header retryAfterHeader) {
+
+		if (retryAfterHeader == null) {
+			return;
+		}
+		String retryAfterAsString = retryAfterHeader.getValue();
+		if (Strings.isNullOrEmpty(retryAfterAsString)) {
+			return;
+		}
+
+		try {
+			DateFormat formatter = new SimpleDateFormat(RESPONSE_RETRY_AFTER_DATE_FORMAT);
+			Date date = formatter.parse(retryAfterAsString);
+
+			Date now = Calendar.getInstance().getTime();
+			long retryAfterAsSeconds = (date.getTime() - convertToDateToGmt(now).getTime()) / 1000;
```

Comment:
Seems like this calculation is always fixed. I might have missed something , please point out to where we are doing exponential backoff in this case.

Reply:
For Throttling requests the server provides the backoff time in the form of the Retry-After header. If that header is missing or invalid (will throw) I call backoff in the catch block

Reply:
Makes sense. I would say to mention a comment!